Add mcxa156 basic environment

NeilChen93 opened the pull request:
Add mcxa156 base enablement

danieldegrasse commented on soc/nxp/mcx/mcxa/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+zephyr_compile_definitions_ifdef(CONFIG_SOC_SERIES_MCXA)
```
This seems like it can be dropped

danieldegrasse commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -45,6 +45,28 @@ static int mcux_lpc_syscon_clock_control_on(const struct device *dev,
+#if defined(CONFIG_SOC_SERIES_MCXA)
+	case MCUX_PORT0_CLK:
+		CLOCK_EnableClock(kCLOCK_GatePORT0);
+		RESET_ReleasePeripheralReset(kPORT0_RST_SHIFT_RSTn);
```
Reset signals should not be handled from the clock control driver. Unfortunately, the best way to handle this is likely to do the following:
- implement a proper reset driver for the MRCC peripheral
- call `reset_line_toggle` from the GPIO driver initialization routine

If you'd like an example for reset drivers, please look here: #72120. I'm aware this is a significant scope creep for enabling this SOC, so as an alternative I am ok if we clear these reset signals in an SOC init function registered with `SYS_INIT` for the time being. I'd prefer we just get the reset driver implemented though

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -321,8 +321,7 @@ static void gpio_mcux_port_isr(const struct device *dev)
 	config->port_base->ISFR = int_status;
-#elif (defined(FSL_FEATURE_GPIO_HAS_INTERRUPT_CHANNEL_SELECT) && \
-		FSL_FEATURE_GPIO_HAS_INTERRUPT_CHANNEL_SELECT)
+#else
```
Was this feature macro dropped in the SDK? If not, why isn't it defined for this GPIO IP block?

danieldegrasse commented on boards/nxp/frdm_mcxa156/Kconfig.defconfig:
Empty files like this can be removed

danieldegrasse reviewed (commented):
A few comments. This PR also requires a change to the west.yml, right? I believe we need to update the revision of the NXP HAL in use to commit 5ed79a9722067 to pull in the SDK 2.16 changes

danieldegrasse commented:
Please don't include merge commits in your PR. Instead, when updating your PR to be based on the latest `main` branch, rebase onto `main` with `git rebase`, then force push your updated branch over the current one

NeilChen93 committed:
dts: mcxa156: add dts for MCXA156

add dts for MCXA156

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

decsny commented on west.yml:
```diff
@@ -198,7 +198,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: e8ccf3e4ad0d6ba80b468747682bde43ec97b435
+      revision: pull/427/head
```
in the manifest commit message, the title should be more specific about what the change in the module is, and the body would be preferable not to describe with "npi" because that is an nxp term, users/contributors of zephyr project might be confused about what it means

decsny commented on soc/nxp/mcx/mcxa/pinctrl_soc.h:
is it possible to use https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h ? it looks pretty much the same

NeilChen93 replied:
Hi @decsny , this file you mentioned is for kinetis project, not suggest for mcxa156.

decsny replied:
the file looks basically the same, why can't you use it for mcxa? is it not the same pin controller as kinetis?

NeilChen93 replied:
Hi @decsny , we don't recommend to use pinctrl_soc_kinetis_common.h for mcxa, mcxa is belong to MCX series, not kinetis. They actually use the same pin controller, but we still hope to make some distinction.

henrikbrixandersen replied:
Could the driver be renamed and reused between the two families?

danieldegrasse replied:
I agree here- I'd like to avoid duplicating things here. If we need to rename the driver to something like `PORT` (what we tend to call this peripheral in the reference manual) then we can, but we should not duplicate implementations

danieldegrasse commented on soc/nxp/mcx/mcxa/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	default 88
+
+config MCUX_LPTMR_TIMER
+	default n if (DT_HAS_NXP_OS_TIMER_ENABLED || DT_HAS_ARM_ARMV8M_SYSTICK_ENABLED)
```
Do we need the `default n` here?

danieldegrasse commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -312,10 +312,12 @@ static int gpio_mcux_manage_callback(const struct device *dev,
 	struct gpio_mcux_data *data = dev->data;
 	uint32_t int_status;
 
+	config = dev->config;
```
Instead of doing this, we could probably use `ARG_UNUSED` macro on these variables, can you try that?

NeilChen93 replied:
Updated, please help review again. Thanks a lot.

danieldegrasse commented on boards/nxp/frdm_mcxa156/frdm_mcxa156.yaml:
```diff
@@ -0,0 +1,19 @@
+name: NXP FRDM MCXA156
+type: mcu
+arch: arm
+ram: 132
```
Does the part have 132KB of ram, or 128KB as the documentation states?

NeilChen93 replied:
Thanks for reminder, update both to 128KB.

decsny reviewed (dismissed):
it appears that the SOC pinctrl_soc.h should be using the common https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h

NeilChen93 committed:
drivers: gpio_mcux: update gpio_mcux driver for MCXA156

Fix compiler warning issue(unused variable 'config',
uninitialized variable 'int_status')

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
drivers: flash: support flash driver for MCXA156

Add MCXA156 romapi support in flash driver

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
boards: frdm_mcxa156: add frdm_mcxa156 board

add frdm_mcxa156 board support

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

decsny commented on soc/nxp/mcx/CMakeLists.txt:
```diff
@@ -5,3 +5,6 @@
 add_subdirectory(${SOC_SERIES})
+
+# This is for access to pinctrl macros
+zephyr_include_directories(../kinetis/common)
```
I don't want to put anything in the family level of MCX still for now (see comments in Kconfig files at family level, I guess I should have put one here as well)

Also, isn't a pinctrl_soc.h file still needed? look at how the kinetis family pinctrl_soc.h is right now, it just includes the common file

NeilChen93 replied:
Updated, thanks a lot. Please help double check.

NeilChen93 committed:
include: pinctrl: Support MCXA in pinctrl_soc_kinetis_common.h

Add bitfield IBE to support MCXA

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
soc: mcxa156: add SOC support for MCXA56

Add MCXA156 support

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

danieldegrasse commented on drivers/clock_control/clock_control_mcux_syscon.c:
Did you run `clang-format` as part of making this change? If so, can you split the change into two commits:
- one commit where you simply run `clang-format` on the files
- one commit where you add the LPUART and PORT clock changes

The reason I ask is because this makes reviewing the PR easier, and will make the git logs easier to follow in the future (particularly `git blame`)

NeilChen93 replied:
Updated, thanks for reminder.

NeilChen93 committed:
drivers: syscon: clang-format syscon driver

clang-format syscon driver

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
drivers: syscon: update syscon driver for MCXA156

Add MCXA port and lpuart clock support

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 commented:
Hi @danieldegrasse @decsny , please help review this PR again, thanks a lot.

mmahadevan108 commented on west.yml:
```diff
@@ -198,7 +198,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: f90b09dea9a7f61e901afa6ce691e687900f5598
```
The SHA should point to commit required by this PR which is 338aec96ae7eace2db263a365ab3fae9bddc48aa

NeilChen93 replied:
Updated.

NeilChen93 committed:
manifest: update HAL NXP manifest

add mcxa156 romapi driver support

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

mmahadevan108 commented:
@decsny @danieldegrasse , can you please re-approve
